network/src/validation.rs

```diff
@@ -505,15 +597,290 @@ impl LiveValidationSessions {
 		where F: FnOnce(Result<&BlockData, Option<&[ValidatorId]>>) -> U
 	{
 		match self.live_instances.get(parent_hash) {
-			Some(c) => c.with_block_data(c_hash, |res| f(res.map_err(Some))),
+			Some(c) => c.1.with_block_data(c_hash, |res| f(res.map_err(Some))),
 			None => f(Err(None))
 		}
 	}
 }
 
+/// Receiver for block data.
+pub struct BlockDataReceiver {
+	outer: Receiver<Receiver<BlockData>>,
+	inner: Option<Receiver<BlockData>>
+}
+
+impl Future for BlockDataReceiver {
+	type Item = BlockData;
+	type Error = io::Error;
+
+	fn poll(&mut self) -> Poll<BlockData, io::Error> {
+		let map_err = |_| io::Error::new(
+			io::ErrorKind::Other,
+			"Sending end of channel hung up",
+		);
+
+		if let Some(ref mut inner) = self.inner {
+			return inner.poll().map_err(map_err);
+		}
+		match self.outer.poll().map_err(map_err)? {
+			Async::Ready(inner) => {
+				self.inner = Some(inner);
+				self.poll()
+			}
+			Async::NotReady => Ok(Async::NotReady),
+		}
+	}
+}
+
+/// Wrapper around bookkeeping for tracking which parachains we're fetching incoming messages
+/// for.
+pub(crate) struct FetchIncoming {
+	exit_signal: ::exit_future::Signal,
+	parachains_fetching: HashMap<ParaId, IncomingReceiver>,
+}
+
+impl FetchIncoming {
+	fn new() -> Self {
+		FetchIncoming {
+			exit_signal: ::exit_future::signal_only(),
+			parachains_fetching: HashMap::new(),
+		}
+	}
+
+	// registers intent to fetch incoming. returns an optional piece of work
+	// that, if some, is needed to be run to completion in order for the future to
+	// resolve.
+	//
+	// impl Future has a bug here where it wrongly assigns a `'static` bound to `M`.
+	fn fetch_with_work<M, W>(&mut self, para_id: ParaId, make_work: M)
+		-> (IncomingReceiver, Option<Box<Future<Item=(),Error=()> + Send>>) where
+		M: FnOnce() -> W,
+		W: Future<Item=Option<Incoming>> + Send + 'static,
+	{
+		let (tx, rx) = match self.parachains_fetching.entry(para_id) {
+			Entry::Occupied(entry) => return (entry.get().clone(), None),
+			Entry::Vacant(entry) => {
+				// has not been requested yet.
+				let (tx, rx) = oneshot::channel();
+				let rx = IncomingReceiver { inner: rx.shared() };
+				entry.insert(rx.clone());
+
+				(tx, rx)
+			}
+		};
+
+		let exit = self.exit_signal.make_exit();
+		let work = make_work()
+			.map(move |incoming| if let Some(i) = incoming { let _ = tx.send(i); })
+			.select2(exit)
+			.then(|_| Ok(()));
+
+		(rx, Some(Box::new(work)))
+	}
+}
+
+/// Can fetch data for a given validation session
+pub struct SessionDataFetcher<P, E, N: NetworkService, T> {
+	network: Arc<N>,
+	api: Arc<P>,
+	fetch_incoming: Arc<Mutex<FetchIncoming>>,
+	exit: E,
+	task_executor: T,
+	knowledge: Arc<Mutex<Knowledge>>,
+	parent_hash: Hash,
+	message_validator: RegisteredMessageValidator,
+}
```

Comment:
Ok so now that I understand the structure a bit better, I have a proposal for a larger restructuring, which perhaps could be done in a separate PR, or here:

I'd like to see if we can remove the two `Arc<Mutex` above, because they potentially introduce blocking in the system(can we know for sure that two tasks will never attempt a `note_statement` or `fetch_incoming` at the same time and end-up having to wait on each other, in the meantime blocking other tasks?).

1. Could we make `SessionDataFetcher` a `Stream`, and only share a sender(probably wrapped in a struct with methods like `fetch_incoming`) with the `Router`? 
2. Then something like `fetch_incoming` could be implemented by having an intermediary  future with an "inner/outer" receiver like we have elsewhere, and sending a message containing a sender to the `SessionDataFetcher` stream, which would then do the work in it's own task and be able to mutate `fetch_incoming` without a lock, finally responding by sending a receiver via the sender it received in the original message.
3. Something like `note_statement` could also be done by sending a message to `SessionDataFetcher`, which would do then do `self.knowledge.note_statement` without having to lock `knowledge`.

Reply:
The inner/outer futures are terrible so I'd prefer to avoid that wherever possible.

Reply:
Some of the operations are order-dependent. e.g. `note_statement` should be done before repropagation. What is the item of `SessionDataFetcher` anyway?

Reply:
I think the item can be `()`, the main idea is that the `SessionDataFetcher` would run in it's own independent task and internally poll channels whose senders have been shared with other tasks, and then "finish" once those senders have been dropped. 

For order dependent operations, you could either do some bookkeeping inside `SessionDataFetcher`(if the order would require cross-task coordination), or if the "order" can be ensured by doing the operations one after the other from one task, you could just send one message after the other from one task and rely on the fact that they will be received in that order. 

I don't think the inner/outer pattern is great, and it does seem like a necessary piece of glue because the other task expect to immediately get a receiver to start polling. It's something that could be addressed if we have a better idea. The good thing about inner/outer is that it cannot block, unlike the use of a lock. 

If you look at `fetch_incoming(&self, parachain: ParaId) -> IncomingReceiver`, you can imagine it running like such:

1. The router calls `self.fetcher.fetch_incoming`, which is supposed to immediately return a receiver.
2. `fetch_incoming` will attempt to acquire the lock around `self.fetch_incoming.lock()`.
3. At that point the thread, not the task, could block waiting for the lock to be available. If this happens it will block every other task tied to the same thread on the pool(`parking_lot` is optimized for threads, not tasks, and will happily park the thread if the lock doesn't become available "quickly"). 
4. When the lock becomes available, the rest of `fetch_incoming` executes and returns a `IncomingReceiver`. 
5. So it looks like we have non-blocking code with the `IncomingReceiver` immediately being available to be included in a chain of work on the thread pool, but actually we have a potential "stop the world" moment right in the middle of the operation. 

Obviously, this can be seen as just an optimization problem, and the lock might actually never be contented, or it can be seen as a code design problem, to be solved by ensuring the code can never block. The ugliness of the inner/outer pattern that might be required to plug this in the system, is actually a result of the system having unrealistic expectations(that a `IncomingReceiver` can always be returned immediately from a call to `fetch_incoming`). 

I don't think it's a huge problem, and perhaps it's best addressed in a follow-up PR. I do think in general it would be good to move away from using `parking_lot` in the context of tasks. If locks really is the right pattern for a given solution, it would be better to see if we could use a lock that is futures aware like https://docs.rs/futures/0.1.25/futures/sync/struct.BiLock.html

Reply:
I've opened https://github.com/paritytech/polkadot/issues/180 to further discuss this...